WIP: Feature/allow ssl

phutchins opened the pull request:
This is still a work in progress. Adding PR for visibility and feedback.

The intent of this PR is to add the ability to enable SSL with regular as well as self signed certificates on the cluster.

Goals:
+ Keep original functionality and backwards compatibility
+ Add a param which would be added to the statefulset to enable SSL
+ By default generate self signed certs for each node
+ Configure mongo when SSL is enabled such that self signed certs work
+ Document the configuration settings that one would use to connect to a cluster that uses self signed certs

Possible Goals:
+ Enable the optional use of hostnames in the replicaset (which would require changes when connecting to the DB from the sidecar as well as when determining the desired replicaset state)

I'm currently running this as it is now in my production Kubernetes environment.

cvallance commented:
@phutchins so far so good, thanks. Let me know when you think it's ready for full review.

cvallance commented:
Hey @phutchins - is this complete? Can we look at testing and merging now?

phutchins commented:
Hey @cvallance, back on this now. Going to look it over and will update here shortly. Thanks for the reminder! 👍

phutchins commented:
@cvallance I'm thinking through the best way to automate(or assist in) the generation and injection of the self signed certificates for mongo. I'd prefer to be able to create and manage/renew them from within the sidecar but pushing to secrets won't work as the stateful set will have already attempted to come up and error'd due to non existent secret, and the mongo containers will also have already attempted to come up without a certificate.

If we could keep the mongo hosts from coming up before the sidecar, we could generate it and either create a secret which we load via a script on the mongo host (gets kinda complicated), or write it to the data mount. I've not run through all of this to see how possible it is however.

Our other option may be to simply include a method in your Makefile that generates the certs and creates a secret. This will be the most straightforward and quickest but the user would have to make sure to renew the certs before they expire...

Maybe we could go the Makefile route and aim for a more automated solution down the road...

Thoughts?

phutchins commented:
I went ahead and added a function to the examples/Makefile to create and save a pem file to a secret. Also made changes to parameterize the default mongod listen port.

cvallance commented:
Awesome. Thanks @phutchins. There's been a few changes to the repo since you last pulled... mongo port configuration has been added and the readme has changed to use tables for settings.

If you pull those changes in and fix things up, I'll merge this.

Thanks again!

cvallance commented on src/lib/config.js:
```diff
@@ -26,6 +26,10 @@ var getKubernetesROServiceAddress = function() {
+  mongoListenPort: process.env.MONGO_LISTEN_PORT || 27017,
+  mongoSSLEnabled: ( process.env.MONGO_SSL_ENABLED === 'true' ) || false,
+  mongoSSLAllowInvalidCertificates: ( process.env.MONGO_SSL_ALLOW_INVALID_CERTIFICATES === 'true' ) || true,
+  mongoSSLAllowInvalidHostnames: ( process.env.MONGO_SSL_ALLOW_INVALID_HOSTNAMES === 'true' ) || true,
```
Also, once you pull, if you have a look at how `isConfigRS` is set for boolean values, use the same approach for these values. Extract it out into a common func too. Cheers.

phutchins replied:
Added but still needs testing.

phutchins commented:
I've rebased from master and have covered most everything I believe. I still need to test it out before we merge...

phutchins commented:
@cvallance I've rebased and tested again and all looks well. I moved the mongoOptions to mongo.js which seems like it goes along with the pattern that already exists more than keeping it in worker.js. Should be good to merge if all still looks good to you.

luqasn commented:
Damn, I didn't see your pull request and added this to my fork as well. I am missing the option to supply a CA certificate and client key in your implementation.

Also, did you test this with mongoSSLAllowInvalidHostnames == false? In my fork, I had to make sure getDb is passed the stable pod hostname (instead of the pod ip, like it is currently implemented) in order for the SSL host verification to work. Plus a workaround to disable ssl validation when connecting to localhost (although disabling host validation should be enough, maybe I should fix that).

See https://github.com/luqasn/mongo-k8s-sidecar/commit/8899d24693bca87bb5a7276e3333227b1c19b0af for my implementation. Currently lacking  documentation and should also add allowInvalidHostnames as option.

luqasn commented on src/lib/mongo.js:
```diff
@@ -172,6 +183,26 @@ var isInReplSet = function(ip, done) {
+// This method will generate a self signed certificate in the format of a PEM file
+// using the addr as the FQDN and the certInfo as the information passed to the certificate
+// (Currently unused)
+var generateCertificate = function(certInfo, done) {
```
Shouldn't we remove this before merging if it is unused?

phutchins replied:
Sure.

luqasn commented on src/lib/mongo.js:
```diff
@@ -2,6 +2,7 @@ var Db = require('mongodb').Db;
 var MongoServer = require('mongodb').Server;
 var async = require('async');
 var config = require('./config');
+var pem = require('pem');
```
Remove if unused

phutchins replied:
I'll move it to a branch and finish this later...

luqasn commented on README.md:
```diff
@@ -96,6 +99,67 @@ mongodb-service-0 # Uses some custom k8s service name. Risks being a duplicate e
+            - "requireSSL"
+            - "--sslPEMKeyFile"
+            - "/data/ssl/mongodb.pem"
+            - "--sslAllowConnectionsWithoutCertificates"
```
I would prefer a safer default without sslAllowConnectionsWithoutCertificates, sslAllowInvalidCertificates and sslAllowInvalidHostnames

phutchins replied:
They default to false...

https://github.com/cvallance/mongo-k8s-sidecar/pull/21/files#diff-832005ddd5bb8e459623773a6c5ffd67R101

... which should be the safest...

The example in the readme is provided for the user to use as needed so its not really a default. We could display/convey this in a different way so that people don't get this as a copy/paste default if it is felt to be necessary.

Its important however as none of this will work without these params without a bit of extra work getting valid ssl certs or a ca updated, etc...

phutchins commented:
@luqasn ah, well hopefully this PR can help you out as well. So I generated the certificates with FQDN of localhost which solves the issue for reaching the nodes for me. That definitely was a bit obscure. I haven't tested with sslAllowInvalidHostnames = false. I'd like to have some tests around all of this eventually...

phutchins commented:
@cvallance I removed the unused code and pushed. This is still ready to go out if you have time to review. Have ideas for next few iterations but want to go ahead and get this merged before I start anything further...

cvallance commented:
Good job, I'm still happy... merging now :shipit:

phutchins commented:
Awesome, thanks @cvallance!
